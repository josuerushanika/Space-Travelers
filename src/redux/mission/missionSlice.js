import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

const url = 'https://api.spacexdata.com/v3/missions';

const initialState = {
  missions: [],
  status: 'idle',
  joined: false,
};

export const fetchMissions = createAsyncThunk('missions/fetchMissions', async () => {
  const response = await fetch(url);
  const data = await response.json();
  return data;
});

const missionSlice = createSlice({
  name: 'missions',
  initialState,
  reducers: {
    joinMission: (state, action) => {
      const missionID = action.payload;
      const updatedMissions = state.missions.map((mission) => {
        if (mission.mission_id === missionID) {
          return { ...mission, joined: true };
        }
        return { ...mission };
      });
      return { ...state, missions: updatedMissions };
    },
    leaveMission: (state, action) => {
      const missionID = action.payload;
      const updatedMissions = state.missions.map((mission) => {
        if (mission.mission_id === missionID) {
          return { ...mission, joined: false };
        }
        return { ...mission };
      });
      return { ...state, missions: updatedMissions };
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchMissions.pending, (state) => ({ ...state, status: 'loading' }))
      .addCase(fetchMissions.fulfilled, (state, action) => ({ ...state, status: 'succeeded', missions: action.payload }))
      .addCase(fetchMissions.rejected, (state) => ({ ...state, status: 'failed' }));
  },
});

export const { joinMission, leaveMission } = missionSlice.actions;
export default missionSlice.reducer;
