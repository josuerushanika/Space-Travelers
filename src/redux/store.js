import { configureStore } from '@reduxjs/toolkit';
import missionReducer from './mission/missionSlice';
import rocketReducer from './rockets/rocketSlice';

const store = configureStore({
  reducer: {
    rockets: rocketReducer,
    missions: missionReducer,
  },
});

export default store;

