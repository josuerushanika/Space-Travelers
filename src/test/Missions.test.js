import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import store from '../redux/store';
import { joinMission } from '../redux/mission/missionSlice';
import Mission from '../components/Missions';

describe('Mission', () => {
  it('renders missions', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <BrowserRouter>
            <Mission />
          </BrowserRouter>
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should Join Mission payload send correct', () => {
    const expectedPayload = { payload: '8HYVG57', type: 'missions/joinMission' };
    const actualPayload = joinMission('8HYVG57');
    expect(actualPayload).toEqual(expectedPayload);
  });
});
