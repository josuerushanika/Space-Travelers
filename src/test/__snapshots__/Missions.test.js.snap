// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mission renders missions 1`] = `
<div>
  Missions
  <table
    className="table table-striped table-bordered table-hover"
  >
    <thead>
      <tr>
        <th>
          Missions
        </th>
        <th>
          Description
        </th>
        <th>
          Status
        </th>
        <th>
           
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;
